src/service_router.rs

```diff
@@ -0,0 +1,90 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use context::{ ContextTrait, Shared };
+use iron::{Request, Response, IronResult};
+use iron::headers::ContentType;
+use iron::status::Status;
+use router::Router;
+use core::marker::Reflect;
+
+
+pub fn generate<Ctx: Send + Reflect + ContextTrait + 'static> (context: Shared<Ctx>) -> Router {
+
+    let mut router = Router::new();
+    let context1 = context.clone();
+    router.get("list.json", move |_: &mut Request| -> IronResult<Response> {
+        // Build a json representation of the services.
+        let ctx = context1.lock().unwrap();
+        let serialized = itry!(ctx.services_as_json());
+
+        let mut response = Response::with(serialized);
+        response.status = Some(Status::Ok);
+        response.headers.set(ContentType::json());
+
+        Ok(response)
+    });
+
+    let context2 = context.clone();
+    router.get(":service", move |req: &mut Request| -> IronResult<Response> {
+        // Call a function on a service.
+        let ctx = context2.lock().unwrap();
+
+        let id = req.extensions.get::<Router>().unwrap().find("service").unwrap_or("");
+        match ctx.get_service(id) {
+            None => {
+                let mut response = Response::with(format!("No Such Service: {}", id));
+                response.status = Some(Status::BadRequest);
+                response.headers.set(ContentType::plaintext());
+                Ok(response)
+            }
+            Some(service) => {
+                service.process_request(req)
+            }
+        }
+    });
+    router
+}
+
+
+extern crate iron_test;
```

Comment:
The compiler is not super happy about it:
```
src/service_router.rs:55:35: 55:42 error: unresolved import `self::iron_test::request`. Could not find `iron_test` in `service_router::service_router` [E0432]
src/service_router.rs:55         pub use self::iron_test::{request, response};
```
I'll make the comment below clearer.